### PR TITLE
[MM-29056] Add Browse Channels, Create New Channel and Open a Direct Message to the category ... menus

### PR DESCRIPTION
**Pull request by AshishDhama**
#### Summary
**Add Browse Channels, Create New Channel and Open a Direct Message to Add Browse Channels, Create New Channel and Open a Direct Message to the category ... menus**

Fixes [https://github.com/mattermost/mattermost-server/issues/17574](https://github.com/mattermost/mattermost-server/issues/17574)

NOTE: Doesn't include "Open a Direct Message" in this PR

#### Ticket Link
[https://mattermost.atlassian.net/browse/MM-29056](https://mattermost.atlassian.net/browse/MM-29056)

#### Related Pull Requests
https://github.com/mattermost/mattermost-server/pull/18396

#### Screenshots

Screenshot 2021-06-14 at 3 40 23 PM
Screenshot 2021-06-14 at 3 47 49 PM
Screenshot 2021-06-14 at 3 39 41 PM

#### Release Note
New features and improvements, including behavioural changes, UI changes and CLI changes

```release-note

1. For each Custom category and the Favorites category, add these new options:
  - Browse Channels
  - Create New Channel
  - Open a Direct Message

2. Once a user adds a channel from this context, that channel should go into the TOP of the category they triggered this menu from.
3. For the Channels category, omit the ‘Open a Direct Message’ option
4. For the Direct Messages category, no changes.
```

### CODE REVIEW

**Comment by AshishDhama** [comment]
Test Snapshot is failing I will update it once someone from the team approves the changes I have done.

**Comment by devinbinnie** [comment]
> Test Snapshot is failing I will update it once someone from the team approves the changes I have done.

Thanks for the PR @AshishDhama, can you update the snapshot before we approve the changes? Thanks!

**Comment by AshishDhama** [comment]
@devinbinnie I have updated the snapshot. I think have used the wrong word "approve" I actually meant if the changes by me are done at the right place and are correct.

File: components/new_channel_flow/new_channel_flow.tsx
Comment (devinbinnie) [inline components/new_channel_flow/new_channel_flow.tsx:140]: Nit: don't need to include a `then()` if we're not doing anything.

**Review by devinbinnie — COMMENTED** [review]
Thanks for the PR @AshishDhama! Code looks good other than my one comment below.

I did notice a couple issues testing:
1) When creating a new channel from that menu, the channel flickers to the "Channels" category and then back to the new one. I think we need to add the channel to the category as we create it. I think there might be an API call around for that so we may need to integrate that.
2) When joining an existing public channel or DM, both of those will go to the default categories instead of the category in which we launched the menu from.

Let us know if you need any help with these issues.

**Comment by AshishDhama** [comment]
@devinbinnie I might need your help as I couldn't find any existing functionality that we can use to create a channel with a category and even the joining function doesn't have any category related param so whether we modify the existing one's or create new to take care of this new scenario please let me know the next step.

**Comment by hmhealey** [comment]
Perhaps we could modify the `createChannel` action to take an optional `categoryId` as an additional parameter and have it add the user to that category instead of calling [`addChannelToInitialCategory`](https://github.com/mattermost/mattermost-webapp/blob/master/packages/mattermost-redux/src/actions/channels.ts#L96) at the end of it? That function only adds the channel to the Channels category visually on the client, so perhaps we can do this in a way that doesn't cause it to flicker into the Channels category?

The only potential problem I can see there is that a websocket event may still cause the channel to be added to Channels before that second call resolves or it may cause it to end up in two categories at once, but I don't know for sure if that would be the case. If that happens, this would likely require changes to the server API, so hopefully we can avoid that.

**Comment by AshishDhama** [comment]
@hmhealey thanks for the suggestion I will try this.

**Comment by hmhealey** [comment]
Cool, let us know if that solves the problem. If not, we'll need to discuss making API changes then which would probably need support from one of the full stack teams.

**Comment by hmhealey** [comment]
As we've discussed one-on-one, we'll need to modify the createChannel API to get the behaviour we want here, so this'll have to be put on hold until that can be done.

File: components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx
Comment (hmhealey) [inline components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx:78]: Minor nit, but could you add a blank line between each of these methods?

File: components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx
Comment (hmhealey) [inline components/sidebar/sidebar_category/sidebar_category_menu/sidebar_category_menu.tsx:301]: I think you might be able to skip the fragment and just return `moreDirectChannelsModal` directly here. There's a chance you might need to set the default to `null` instead `undefined`, but I don't know if that would also be needed.

**Review by hmhealey — APPROVED** [review]
A couple non-blocking comments, but this otherwise looks good!

I'm looking forward to this since I've actually been trying to access Browse Channels from the category menu since the new sidebar went in around last December, so it'll be nice that it's going to be where I keep expecting it to be

**Comment by AshishDhama** [comment]
@esethna @anurag6713 was working on this as part of the productivity sprint. I think there are minor changes required to finish this up.

**Comment by anurag6713** [comment]
Yes I would like to finish the third use case as well with the DMs and GMs. That's why I haven't merged it yet even though I initially said I would skip it 😅

**Comment by Willyfrog** [comment]
@AshishDhama @anurag6713 what is missing on this PR to move it forward? other than making it up to date

**Comment by anurag6713** [comment]
"Open a direct message" part is still pending. Needs changes on server PR as well. I will take this up on Fridays after the alpha.

**Comment by esethna** [comment]
@AshishDhama just checking on this if you or @anurag6713 are still working on it?

**Comment by anurag6713** [comment]
I will continue working on this from this Friday as I am almost done with mobile alpha 🙂

**Comment by Willyfrog** [comment]
@anurag6713 I see that this is still pending, regardging

> "Open a direct message" part is still pending. Needs changes on server PR as well. I will take this up on Friday after the alpha.

maybe we could split that into a separate PR so we can merge this one?

**Comment by AshishDhama** [comment]
Closing this one, if required will open it again.

**Pull request closed without merging**
